File: file/src/main/scala/farjs/file/popups/FileViewHistoryController.scala
package farjs.file.popups

import farjs.file.FileEvent
import farjs.filelist.stack.{WithStacks, WithStacksProps}
import scommons.react._

import scala.scalajs.js

object FileViewHistoryController extends FunctionComponent[FileViewHistoryControllerProps] {

  private[popups] var fileViewHistoryPopup: UiComponent[FileViewHistoryPopupProps] =
    FileViewHistoryPopup

  protected def render(compProps: Props): ReactElement = {
    val stacks = WithStacks.useStacks()
    val props = compProps.plain

    if (props.showPopup) {
      <(fileViewHistoryPopup())(^.plain := FileViewHistoryPopupProps(
        onAction = { history =>
          props.onClose()

          WithStacksProps.active(stacks).input.emit("keypress", js.undefined, js.Dynamic.literal(
            name = "",
            full = FileEvent.onFileView,
            data = history.asInstanceOf[js.Dynamic]
          ))
        },
        onClose = props.onClose
      ))()
    }
    else null
  }
}

File: file/popups/FileViewHistoryController.mjs
import React from "react";
import WithStacks from "@farjs/filelist/stack/WithStacks.mjs";
import WithStacksProps from "@farjs/filelist/stack/WithStacksProps.mjs";
import FileViewHistoryPopup from "./FileViewHistoryPopup.mjs";
import FileEvent from "../FileEvent.mjs";

const h = React.createElement;

/**
 * @typedef {{
 *  readonly showPopup: boolean;
 *  onClose(): void;
 * }} FileViewHistoryControllerProps
 */

/**
 * @param {FileViewHistoryControllerProps} props
 */
const FileViewHistoryController = (props) => {
  const { fileViewHistoryPopup } = FileViewHistoryController;

  const stacks = WithStacks.useStacks();

  return props.showPopup
    ? h(fileViewHistoryPopup, {
        onAction: (history) => {
          props.onClose();

          WithStacksProps.active(stacks).input.emit("keypress", undefined, {
            name: "",
            full: FileEvent.onFileView,
            data: history,
          });
        },
        onClose: props.onClose,
      })
    : null;
};

FileViewHistoryController.displayName = "FileViewHistoryController";
FileViewHistoryController.fileViewHistoryPopup = FileViewHistoryPopup;

export default FileViewHistoryController;
